### PR TITLE
Aleem/mentor

### PR DESCRIPTION
**Pull request by AleemKhan12**
Moved some shared interfaces and functions to utils folder and created mentor ui pages

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -21,6 +21,7 @@
     "eslint-config-react-app": "^7.0.1",
     "jest-watch-typeahead": "^1.0.0",
     "luxon": "^2.3.2",
+    "moment": "^2.29.3",
```
Comment (wley3337) [inline package.json:24]: I think we want to use `luxon` for moment type stuff. Moment isn't being maintained :(

File: src/components/mentor/MentorBooking.tsx [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+interface EventComponenetProps {
+    event: CalendarEvent; 
+    title?:string;
+  }
+  
+
+const localizer = luxonLocalizer(DateTime);
+
+////Custom Components 
+const customComponents = {
```
Comment (wley3337) [inline src/components/mentor/MentorBooking.tsx:22]: Could we rename this to something like: `calendarEventComponents` or `calendarComponents` so it's clear what it's for?

File: src/components/mentor/MentorBookingModal.tsx [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+  
+
+    const isEvent = isCalendarEvent(eventOrSlot);
+    const [disable, setDisable] = useState<boolean>(true);
+
+    return (
+      <Modal
+        open={Boolean(eventOrSlot)}
+        onClose={onClose}
+        // @ts-ignore
```
Comment (wley3337) [inline src/components/mentor/MentorBookingModal.tsx:44]: This makes me nervous :) Why do we need ts to ignore this?

File: src/components/mentor/MentorBookingModal.tsx [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        }
+        actions={
+          <Box display="flex" justifyContent="flex-end" width="100%">
+            <Button onSubmit={() => onClose()} disabled={disable}> {isEvent ? 'Book' : 'Create'}</Button>
+          </Box>
+        }
+      >
+        <Grid container spacing={2} py={2}>
+          <Grid item xs={12}>
+            <Autocomplete<string, false, false, false>
```
Comment (wley3337) [inline src/components/mentor/MentorBookingModal.tsx:58]: This syntax feels off to me 🤔

File: src/components/mentor/MentorBookingModal.tsx [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+              )}
+            />
+          </Grid>
+          <Grid item xs={12}>
+            <TextField
+              fullWidth
+              label="Date"
+              placeholder={isEvent ? eventOrSlot.start.toString() : ''}
+              value={isEvent ? eventOrSlot.start.toString() : ''}
+              InputProps={{ startAdornment: <AccessTime /> }}
```
Comment (wley3337) [inline src/components/mentor/MentorBookingModal.tsx:81]: I don't understand why we're passing a component down as props here. Is this the pattern we should be following?

File: src/components/mentor/MentorBookingModal.tsx [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    const isEvent = isCalendarEvent(eventOrSlot);
+    const [disable, setDisable] = useState<boolean>(true);
+
+    return (
+      <Modal
+        open={Boolean(eventOrSlot)}
+        onClose={onClose}
+        // @ts-ignore
+        title={
+          <Box display="flex" alignItems="center">
```
Comment (wley3337) [inline src/components/mentor/MentorBookingModal.tsx:46]: This is of a type `string` as far as prop types. I wonder if this is what TS is complaining about. If we need CSS passed down, I think we may want to add a style prop to the Modal or just make the `tile: string | React.Node` kind of thing

File: src/components/mentor/MentorHome.tsx [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+import { useMemo, useState } from 'react';
+import MentorSessionCard from './MentorSessionCard';
+import { getMentorProfile } from '../../utils/api';
+
+const EmptySessionMsg = () => {
+
+    return (
+        <Grid item xs={12} pb={2}>
+            <Typography variant="body1">
+                You haven't booked any sessions yet. <Link to='/student/booking'>Book</Link> your sessions now!
```
Comment (wley3337) [inline src/components/mentor/MentorHome.tsx:13]: should this be `student/booking` or should it just be `/booking` and the user context determines if it's student or mentor?

File: src/components/mentor/MentorHome.tsx [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+            </Typography>
+        </Grid>
+    );
+};
+
+
+const MentorHome = () => {
+
+    const [checkIn, setCheckIn] = useState<boolean | null>(null);
+    const user = useMemo(()=> getMentorProfile('1'), []);
```
Comment (wley3337) [inline src/components/mentor/MentorHome.tsx:23]: So I'm not familiar with the `useMemo` hook and I _think_ you can use it in combination with `useSelector`. I don't know if you'd see gains in using it in this context.
Ultimately this would want to be a `useSelector(state=> state.user.mentorProfile)` here. If that feels a bit wordy to write each time, we could make a selector ( like an action creator ) for it and put it under a `redux/user/user.selectors.ts`

File: src/components/mentor/MentorIndex.tsx [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+const menuItems = [
+    { label: 'Home', url: '/mentor' },
+    { label: 'Booking', url: '/mentor/booking' },
+    { label: 'My Sessions', url: '/mentor/session' },
+    { label: 'Messages', url: '/mentor/message' },
+  ];
+  
+
+const MentorIndex = () => {
```
Comment (wley3337) [inline src/components/mentor/MentorIndex.tsx:17]: Do we want to establish a naming convention here for the internal routing? Not sure if we want to go Index, Controller, Router 🤷

File: src/utils/api.ts [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+export const getStudentProfile = (id?:string):Profile | undefined => {
+    if (id == undefined) return;
+    return studentProfiles.filter(profile => profile.id == id)[0];
+}
+
+//--------------------------------------------------------------------------
+
+//Announcements
+export const getAnnounements = (program?: string):Announcement[] => {
```
Comment (wley3337) [inline src/utils/api.ts:46]: It make sense to me that an Announcement would need to be called with an id for the thing you want the announcements of. I totally forgot that this was a one to many relationship as well, so I'll add the Announcement model and include them as part of the library serializer 🤦

File: yarn.lock [hunk 11]
```diff
@@ -8662,6 +8662,13 @@ __metadata:
   languageName: node
   linkType: hard
 
+"moment@npm:^2.29.3":
```
Comment (wley3337) [inline yarn.lock:8665]: Just a reminder not to try and edit this file when you pull moment out. Just delete it and auto generate it again with `yarn` Not that I've ever made that mistake and broken my project before.... never.

**Comment by wley3337** [comment]
Oh, don't forget to run prettier against all your files too :)

File: src/components/mentor/MentorBookingModal.tsx (hunk 4 above)
Comment (geekingreen) [reply src/components/mentor/MentorBookingModal.tsx:58]: Funny enough that is the correct syntax for passing types to a component and in particular this component has some boolean values as well, which can potentially be excluded here if false is the default. `<TypeOfSearchable, Multiple, DisableClearable, FreeSolo>` are the values it is looking for, only the first is required unless props to change the three others have also been passed in

File: src/components/mentor/MentorBookingModal.tsx (hunk 5 above)
Comment (geekingreen) [reply src/components/mentor/MentorBookingModal.tsx:81]: This also is the correct usage of adding an adornment to a mui `TextField`

Comment (wley3337) [reply src/components/mentor/MentorBookingModal.tsx:81]: Thank you!

File: src/components/mentor/MentorBookingModal.tsx (hunk 4 above)
Comment (wley3337) [reply src/components/mentor/MentorBookingModal.tsx:58]: Cool! I've never seen types passed to a component before!

File: src/components/mentor/MentorBookingModal.tsx (hunk 6 above)
Comment (geekingreen) [reply src/components/mentor/MentorBookingModal.tsx:46]: I made the `Modal` component in hopes that we wouldn't need to style it unless we change it as a whole. My hope is to make it as simple as possible to use, and hopefully have the styles of all modals match. So perhaps if we want titles to be centered we modify `Modal` so that all modals have centered titles?

File: yarn.lock (hunk 11 above)
Comment (geekingreen) [reply yarn.lock:8665]: another option is to use `yarn remove <packagename>` or in this particular instance `yarn remove moment` it will remove the entry from `package.json` and update `yarn.lock` all in one go

File: src/components/mentor/MentorHome.tsx (hunk 8 above)
Comment (wley3337) [reply src/components/mentor/MentorHome.tsx:23]: So it looks like [useMemo docs](https://reactjs.org/docs/hooks-reference.html#usememo) say it's used to reduce the cost of expensive tasks. I think we may want to drop it here since we're not fetching the user we're just reading it from the store

**Review by geekingreen — CHANGES_REQUESTED** [review]
Lookin' good, I agree with @wley3337's comments (except as noted) especially running `prettier` 😄

File: src/components/mentor/MentorHome.tsx (hunk 7 above)
Comment (geekingreen) [reply src/components/mentor/MentorHome.tsx]: I think this is okay here since we know our context is Mentor?

File: src/components/mentor/MentorIndex.tsx (hunk 9 above)
Comment (geekingreen) [reply src/components/mentor/MentorIndex.tsx:16]: I think for each main route, `/student` `/mentor` `/advisor` `/librarian`, etc... we can use `<Name>Index` (which I think we've done so far) for the sub routes, I haven't kept up a naming convention... but we could append something to make it obvious in the filename without having to dig into the `Index`'s like `<AdvisorComputersRoute />` or even just dump them into a folder called `routes` within it's context... otherwise as long as we have readable code from the index routes it should be easy enough to see what's going on

**Comment by geekingreen** [comment]
@wley3337 cleaned up things a bit, let me know if there is anything else that could use a touchup. If we want to modify our convention for routes, I think we should do that in a separate PR.
